lib/brainstem/api_docs/formatters/markdown/endpoint_formatter.rb

```diff
@@ -83,59 +83,63 @@ def format_endpoint!
-                    root_param_name,
-                    endpoint.valid_params[root_param_name],
-                    0
-                  )
-                else
-                  text = md_inline_code(root_param_name) + "\n"
-
-                  child_keys.each do |param_name|
-                    text += parameter_with_indent_level(
-                      param_name,
-                      endpoint.valid_params[param_name],
-                      1
-                    )
-                  end
-
-                  buff << md_li(text)
-                end
-
-                buff
+              params_configuration_tree.inject("") do |buff, (field_name, field_info)|
+                buff << format_param_tree!("", field_name, field_info)
               end
             end
           end
 
 
+          # Formats the parent parameter and its children
+          def format_param_tree!(buffer, field_name, field_info, indentation = 0)
+            buffer += parameter_with_indent_level(
+              field_name,
+              field_info,
+              indentation
+            )
+
+            children = field_info['children'] || []
+            children.each do |child_field_name, child_field_info|
+              buffer = format_param_tree!(buffer, child_field_name, child_field_info, indentation + 1)
+            end
+
+            buffer
+          end
+
           #
           # Formats a given parameter with a variable indent level. Useful for
           # indifferently formatting root / nested parameters.
           #
           # @param [String] name the param name
           # @param [Hash] options information pertinent to the param
+          # @option [Boolean] options :required
           # @option [Boolean] options :legacy
           # @option [Boolean] options :recursive
           # @option [String,Symbol] options :only Deprecated: use +actions+
           #   block instead
           # @option [String] options :info the doc string for the param
+          # @option [String] options :type The type of the field.
+          #   e.g. string, integer, boolean, array, hash
+          # @option [String] options :item The type of the items in the field.
+          #   Ideally used when the type of the field is an array or hash.
           # @param [Integer] indent how many levels the output should be
           #   indented from normal
           #
           def parameter_with_indent_level(title, options = {}, indent = 0)
-            options = options.dup
+            options = options.deep_dup
             text    = md_inline_code(title)
 
+            text += md_inline_type(options.delete(:type), options.delete(:item)) if options.has_key?(:type)
             text += " - #{options.delete(:info)}" if options.has_key?(:info)
 
             if options.keys.any?
               text += "\n"
+              text += md_li("Required: #{options[:required].to_s}",   indent + 1) if options.has_key?(:required) && options[:required]
```

Comment:
Opinionated: I would pull out `indent + 1` and name it accordingly. It seems that it will always be the next indent.

Reply:
Perhaps the indent parameter should already be incremented? It looks a bit inconsistent in `format_param_tree`